extensions/machine-learning-services/src/packageManagement/sqlPackageManageProvider.ts

```diff
@@ -0,0 +1,181 @@
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as vscode from 'vscode';
+import * as azdata from 'azdata';
+import * as nbExtensionApis from '../typings/notebookServices';
+import * as utils from '../common/utils';
+import * as constants from '../common/constants';
+import { QueryRunner } from '../common/queryRunner';
+import { ApiWrapper } from '../common/apiWrapper';
+import { ProcessService } from '../common/processService';
+
+const installMode = 'install';
+const uninstallMode = 'uninstall';
+const localPythonProviderId = 'localhost_Pip';
+
+/**
+ * Manage Package Provider for python packages inside SQL server databases
+ */
+export class SqlPythonPackageManageProvider implements nbExtensionApis.IPackageManageProvider {
+
+	private _pythonExecutable: string;
+
+	public static ProviderId = 'sql_Python';
+
+	/**
+	 * Creates new a instance
+	 */
+	constructor(
+		private _nbExtensionApis: nbExtensionApis.IExtensionApi,
+		private _outputChannel: vscode.OutputChannel,
+		private _rootFolder: string,
+		private _apiWrapper: ApiWrapper,
+		private _queryRunner: QueryRunner,
+		private _processService: ProcessService) {
+		this._pythonExecutable = utils.getPythonExePath(this._rootFolder);
+	}
+
+	/**
+	 * Returns provider Id
+	 */
+	public get providerId(): string {
+		return SqlPythonPackageManageProvider.ProviderId;
+	}
+
+	/**
+	 * Returns package target
+	 */
+	public get packageTarget(): nbExtensionApis.IPackageTarget {
+		return { location: 'SQL', packageType: 'Python' };
+	}
+
+	/**
+	 * Returns list of packages
+	 */
+	public async listPackages(): Promise<nbExtensionApis.IPackageDetails[]> {
+		let packages = await this._queryRunner.getPythonPackages(await this.getCurrentConnection());
+		if (packages) {
+			packages = packages.sort((a, b) => a.name.localeCompare(b.name));
+		} else {
+			packages = [];
+		}
+		return packages;
+	}
+
+	/**
+	 * Installs given packages
+	 * @param packages Packages to install
+	 * @param useMinVersion minimum version
+	 */
+	async installPackages(packages: nbExtensionApis.IPackageDetails[], useMinVersion: boolean): Promise<void> {
+		if (packages) {
+			for (let index = 0; index < packages.length; index++) {
+				const element = packages[index];
+				await this.updatePackage(element, installMode);
+			}
+		}
+		//TODO: use useMinVersion
+		console.log(useMinVersion);
+	}
+
+	/**
+	 * Execute a script to install or uninstall a python package inside current SQL Server connection
+	 * @param packageDetails Packages to install or uninstall
+	 * @param scriptMode can be 'install' or 'uninstall'
+	 */
+	private async updatePackage(packageDetails: nbExtensionApis.IPackageDetails, scriptMode: string): Promise<void> {
+		let connection = await this.getCurrentConnection();
+		let credentials = await this._apiWrapper.getCredentials(connection.connectionId);
+
+		if (connection) {
+			let port = '1433';
```

Comment:
How do you know the port is always going to be 1433?

Reply:
default is 1433 unless it's included in server name 

---
In reply to: [358523634](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r358523634) [](ancestors = 358523634)